include/revng/ADT/Concepts.h

```diff
@@ -4,44 +4,178 @@
+constexpr auto checkAllTupleElementTypes() {
+  constexpr std::size_t Size = std::tuple_size_v<T>;
+  return checkTupleElementTypes<T>(std::make_index_sequence<Size>());
+}
+
+} // namespace revng::detail
+
+// clang-format off
+template<class T>
+concept TupleLike = (TupleSizeCompatible<T>
+                     and revng::detail::checkAllTupleElementTypes<T>());
+// clang-format on
+
+static_assert(TupleLike<std::tuple<>>);
+static_assert(TupleLike<std::tuple<int, int, long>>);
+static_assert(TupleLike<std::pair<int, int>>);
+static_assert(TupleLike<std::array<int, 0>>);
+static_assert(not TupleLike<std::vector<int>>);
+
+//
+// Concepts to simplify working with specializations of templates.
+//
+
+/// A concept that helps determine whether a given object is a specialization
+/// (or is inheriting a specialization) of a given template.
+///
+/// This lets us cut down on the number of `IsX` (e.g. `IsGenericGraph`,
+/// `IsRank`) concepts needed since we can now just say
+/// `SpecializationOf<GenericGraph>` on the interface boundaries and get
+/// the expected check.
+///
+/// TODO: this requires clang-13+, uncomment it after the update.
+// template<typename Type, template<typename...> class Ref>
+// concept SpecializationOf = requires(Type &&Value) {
+//   []<typename... Ts>(Ref<Ts...> &){}(Value);
+// };
+//
+// template<typename Type, template<typename...> class Ref>
+// concept NotSpecializationOf = !SpecializationOf<Type, Ref>;
+//
+// static_assert(SpecializationOf<std::pair<int, long>, std::pair>);
+// static_assert(SpecializationOf<const std::pair<int, long>, std::pair>);
+// static_assert(SpecializationOf<std::string, std::basic_string>);
+// static_assert(SpecializationOf<std::ostream, std::basic_ios>);
+// static_assert(not SpecializationOf<std::string, std::basic_string_view>);
+
+/// TODO: Remove after updating to clang-format with concept support.
+struct ClangFormatPleaseDoNotBreakMyCode;
+// clang-format off
+// clang-format on
+
+namespace revng::detail {
+
+template<typename Test, template<typename...> class Ref>
+struct StrictSpecializationHelper : std::false_type {};
+
+template<template<typename...> class Ref, typename... Args>
+struct StrictSpecializationHelper<Ref<Args...>, Ref> : std::true_type {};
+
+template<template<typename...> class Ref, typename... Args>
+struct StrictSpecializationHelper<const Ref<Args...>, Ref> : std::true_type {};
+
+template<typename Test, template<typename...> class Ref>
+constexpr bool
+  StrictSpecialization = StrictSpecializationHelper<Test, Ref>::value;
+
+} // namespace revng::detail
+
+/// A more strict version of the `SpecializationOf` concept. This one only
+/// allows direct specializations (and aliases) - no inheritance.
+///
+/// It's useful in the cases when template parameter deduction is important,
+/// e.g. when instanciating traits.
+template<typename Test, template<typename...> class Ref>
+concept StrictSpecializationOf = revng::detail::StrictSpecialization<Test, Ref>;
```

Comment:
The `Strict` prefix was to distinguish it from places where we use were willing to accept derived classes to, but here it seems you don't even have a *non-*Strict version of the concept, so I think we can strip the prefix and be happy.

Reply:
Actually, I'd prefer to keep it and reintroduce a "non-strict" version as soon as we are able (updating to clang13+ will do the trick)